### PR TITLE
Modernize the interestingness tests

### PR DESCRIPTION
**Pull request by pyoor**

### CODE REVIEW

File: src/lithium/interestingness/timed_run.py [hunk 1]
```diff
@@ -132,46 +184,41 @@ def timed_run(
...
+        status = ExitStatus.ABNORMAL
     else:
-        # return_code < 0 (or > 0x80000000 in Windows)
-        # The program was terminated by a signal, which usually indicates a crash.
+        # The program was terminated by a signal or by the sanitizer (ERROR_CODE)
         # Mac/Linux only!
-        # XXX: this doesn't work on Windows
         if child.returncode < 0:
-            signum = -child.returncode
+            signum = abs(child.returncode)
```
Comment (jschwartzentruber) [inline src/lithium/interestingness/timed_run.py:209]: Why switch this? Either way works, just curious.

File: tests/test_interestingness.py [hunk 2]
```diff
@@ -154,10 +154,8 @@ def test_diff_test_0() -> None:
             "diff_test",
             "--timeout",
             "99",
-            "-a",
-            "flags_one",
-            "-b",
-            "flags_two_a flags_two_b",
+            "-a 'flags_one'",
+            "-b 'flags_two_a flags_two_b'",
```
Comment (jschwartzentruber) [inline tests/test_interestingness.py:158]: What does this test do? These args look wrong.

**Review by jschwartzentruber — CHANGES_REQUESTED** [review]

File: src/lithium/interestingness/timed_run.py (hunk 1 above)
Comment (pyoor) [reply src/lithium/interestingness/timed_run.py:209]: abs() just seems more explicit to me.

File: pyproject.toml [hunk 3]
```diff
@@ -34,6 +34,8 @@ disable = [
     "duplicate-code",
     "fixme",
     "import-error",
+    "logging-fstring-interpolation",
+    "missing-module-docstring",
```
[the remark this answers is not in the document]
Comment (pyoor) [reply pyproject.toml]: Fixed via [86da624](https://github.com/MozillaSecurity/lithium/pull/92/commits/86da6248fc3aab0086d115a18ebc012609011ab1).

File: src/lithium/interestingness/diff_test.py [hunk 4]
```diff
@@ -17,80 +16,106 @@
...
             return True
     else:
-        log.info("[Uninteresting] At least one test timed out. %s", time_str)
-        return False
+        if cmp_out(a_run.out, b_run.out) or cmp_out(a_run.err, b_run.err):
+            LOG.info("Differences in output detected")
+            return True
 
-    log.info("[Uninteresting] Identical behaviour. %s", time_str)
+    LOG.info("No differences detected")
```
[the remark this answers is not in the document]
Comment (pyoor) [reply src/lithium/interestingness/diff_test.py]: Fixed via [aa83936](https://github.com/MozillaSecurity/lithium/pull/92/commits/aa839365d46fc83795fb16ea6b894d0ba4cdf7b6).

File: src/lithium/interestingness/crashes.py [hunk 5]
```diff
@@ -25,20 +23,18 @@ def interesting(cli_args: List[str], temp_prefix: str) -> bool:
...
-
-    time_str = f" ({runinfo.elapsedtime:.3f} seconds)"
-    if runinfo.sta == timed_run.CRASHED:
-        log.info("Exit status: " + runinfo.msg + time_str)
+    if run_info.status == ExitStatus.CRASH:
+        LOG.info(f"Crash detected ({run_info.elapsed:.3f}s)")
         return True
 
-    log.info("[Uninteresting] It didn't crash: " + runinfo.msg + time_str)
+    LOG.info(f"No crash detected ({run_info.elapsed:.3f}s)")
```
[the remark this answers is not in the document]
Comment (pyoor) [reply src/lithium/interestingness/crashes.py]: Fixed via [aa83936](https://github.com/MozillaSecurity/lithium/pull/92/commits/aa839365d46fc83795fb16ea6b894d0ba4cdf7b6).

File: src/lithium/interestingness/diff_test.py [hunk 6]
```diff
@@ -17,80 +16,106 @@
...
+
+    if temp_prefix:
+        if cmp_out(a_run.out, b_run.out, True) or cmp_out(a_run.err, b_run.err, True):
+            LOG.info("Differences in output detected")
             return True
     else:
-        log.info("[Uninteresting] At least one test timed out. %s", time_str)
-        return False
+        if cmp_out(a_run.out, b_run.out) or cmp_out(a_run.err, b_run.err):
+            LOG.info("Differences in output detected")
```
[the remark this answers is not in the document]
Comment (pyoor) [reply src/lithium/interestingness/diff_test.py]: Fixed via [aa83936](https://github.com/MozillaSecurity/lithium/pull/92/commits/aa839365d46fc83795fb16ea6b894d0ba4cdf7b6).

File: src/lithium/interestingness/diff_test.py [hunk 7]
```diff
@@ -17,80 +16,106 @@
...
+        b_run: Union[str, bytes],
+        is_file: bool = False,
+    ) -> bool:
+        if is_file:
+            return not filecmp.cmp(a_data, b_run)
+        return a_data != b_run
+
+    if temp_prefix:
+        if cmp_out(a_run.out, b_run.out, True) or cmp_out(a_run.err, b_run.err, True):
+            LOG.info("Differences in output detected")
```
[the remark this answers is not in the document]
Comment (pyoor) [reply src/lithium/interestingness/diff_test.py]: Fixed via [aa83936](https://github.com/MozillaSecurity/lithium/pull/92/commits/aa839365d46fc83795fb16ea6b894d0ba4cdf7b6).

File: src/lithium/interestingness/crashes.py [hunk 8]
```diff
@@ -1,22 +1,20 @@
...
-Example:
-    python -m lithium crashes --timeout=9 <binary> --fuzzing-safe <testcase>
-"""
-
 import logging
-from typing import List
+import sys
+from typing import List, Optional
+
+from lithium.interestingness.timed_run import BaseParser, ExitStatus, timed_run
```
[the remark this answers is not in the document]
Comment (pyoor) [reply src/lithium/interestingness/crashes.py]: Fixed via [04e7e87](https://github.com/MozillaSecurity/lithium/pull/92/commits/04e7e870b3ed0715862cd5782c308e953051cf00).

**Review by jschwartzentruber — APPROVED** [review]

File: src/lithium/interestingness/diff_test.py [hunk 9]
```diff
@@ -17,80 +16,106 @@
...
+    # Run with arguments set B
+    command_b = binary + args.b_args.split() + testcase
+    log_prefix_b = f"{temp_prefix}-b" if temp_prefix else None
+    b_run = timed_run(command_b, args.timeout, log_prefix_b)
+    if b_run.status == ExitStatus.TIMEOUT:
+        LOG.warning("Command B timed out!")
+
+    # Compare return codes
+    if a_run.return_code != b_run.return_code:
+        LOG.info(f"Different return codes: {a_run.return_code} vs {b_run.return_code}")
```
[the remark this answers is not in the document]
Comment (pyoor) [reply src/lithium/interestingness/diff_test.py]: Fixed via [aa83936](https://github.com/MozillaSecurity/lithium/pull/92/commits/aa839365d46fc83795fb16ea6b894d0ba4cdf7b6).

File: tests/test_interestingness.py (hunk 2 above)
Comment (pyoor) [reply tests/test_interestingness.py]: Fixed via [c6b41e7](https://github.com/MozillaSecurity/lithium/pull/92/commits/c6b41e7d573e772fcd933d66ba47c4f2b54793f6).

**Pull request merged**